### PR TITLE
fixed SINGULARITYENV_PATH and added SINGULARITYENV_APPEND_PATH and SINGULARITYENV_PREPEND_PATH

### LINKED ISSUE #1352: SINGULARITYENV_PATH doesn't appear to be propagated into the runtime container environment
## Version of Singularity:

 2.4.2

### Expected behavior

It should be possible to execute programs located in the path set via SINGULARITYENV_PATH variable.

### Actual behavior

Debug log shows the SINGULARITYENV_PATH is being processed into PATH, but the resulting PATH doesn't appear to apply. jms replicated the issue.

### Steps to reproduce behavior

Put a static binary into a sandbox container, say /opt/foo/myprog

Export SINGULARITYENV_PATH=/opt/foo before executing the container with 'myprog'.

The following error is shown:

/.singularity.d/actions/exec: 9: exec: prog: not found

If the correct PATH is exported from /.singularity.d/env/90-environment.sh 'prog' will be found and executed.

### PR DESCRIPTION
**Pull request by GodloveD**
**Description of the Pull Request (PR):**

In the days of yore, a user could specify a value of `SINGULARITYENV_PATH` and this would be prepended to the `PATH` inside their container at runtime.  Some time ago, this option was broken by the addition of a file in `/.singularity.d/env` called `10-docker.sh` which overwrites the base path set in c with whatever path is set in Docker.  This is necessary because the user may have added custom paths to the container in the Docker file.  But it breaks `SINGULARITYENV_PATH`.

This PR will fix the problem for newly built containers and also give users the ability to overwrite and append to the container `PATH` at runtime with environment variables.  At build time, a new file will be added to the `/.singularity.d/env` directory called `99-runtimevars.sh`.  When running one of these newly built containers, the user can set these variables for the corresponding `PATH` adjustments.  

- `SINGULARITYENV_PATH` <- overwrite the container `PATH` at runtime.  (Note that this changes the old behavior which was to prepend the contents of the variable onto `PATH`.
- `SINGULARITYENV_PREPEND_PATH` <- prepend some new content to the container `PATH` at runtime
- `SINGULARITYENV_APPEND_PATH` <- append some new content to the container `PATH` at runtime

**This fixes or addresses the following GitHub issues:**

fixes #1352
see also #1363 and #1389 

**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [ ] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge

Attn: @singularityware-admin

### CODE REVIEW

**Comment by dtrudg** [comment]
This is giving correct behaviour for me.

**No Dockerfile custom path, No SINGULARITYENV_PATH**

```
singularity exec docker://alpine env | grep ^PATH=
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**No Dockerfile custom path, Set SINGULARITYENV_PATH**
```
SINGULARITYENV_PATH=/usr/bin:/bin singularity exec docker://alpine env | grep ^PATH= 
PATH=/usr/bin:/bin
```

**No Dockerfile custom path, Set SINGULARITYENV_APPEND_PATH**
```
SINGULARITYENV_APPEND_PATH=/opt singularity exec docker://alpine env | grep ^PATH= 
PATH=/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt
```

**No Dockerfile custom path, Set SINGULARITYENV_PREPEND_PATH**
```
SINGULARITYENV_PREPEND_PATH=/opt singularity exec docker://alpine env | grep ^PATH= 
PATH=/opt:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**Dockerfile custom path, No SINGULARITYENV_PATH**
```
singularity exec docker://continuumio/miniconda env | grep ^PATH= 
PATH=/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

**Dockerfile custom path, Set SINGULARITYENV_PATH**
```
SINGULARITYENV_PATH=/usr/bin:/bin singularity exec docker://continuumio/miniconda env | grep ^PATH=
PATH=/usr/bin:/bin
```

**Dockerfile custom path, Set SINGULARITYENV_APPEND_PATH**
```
SINGULARITYENV_APPEND_PATH=/testpath singularity exec docker://continuumio/miniconda env | grep ^PATH=
PATH=/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/testpath
```

**Dockerfile custom path, Set SINGULARITYENV_PREPEND_PATH**
```
SINGULARITYENV_PREPEND_PATH=/testpath singularity exec docker://continuumio/miniconda env | grep ^PATH=
PATH=/testpath:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

File: etc/init [hunk 1]
```diff
@@ -10,12 +10,28 @@ unset ml
...
 unset BASH_ENV
 
-# Provide a sane path within the container
-if [ -z "${SINGULARITYENV_PATH:-}" ]; then
-    SINGULARITYENV_PATH="/bin:/sbin:/usr/bin:/usr/sbin:/usr/local/bin:/usr/local/sbin"
-else
-    SINGULARITYENV_PATH="$SINGULARITYENV_PATH:/usr/local/bin:/usr/local/sbin:/bin:/sbin:/usr/bin:/usr/sbin"
-fi
+# These vars are provided to the user as a convenience.  They will be used to 
+# prepended, appended, or overwrite the PATH at runtime.
```
Comment (dtrudg) [inline etc/init]: `prepend to, append to, or overwrite PATH at runtime`

**Review by dtrudg — CHANGES_REQUESTED** [review]
This is all working for me, but I think it is important this correct behaviour should really be captured in the test suite so we can guard against regressions going forward. Maybe the run commands in comment above give some test examples that can be worked from?

**Comment by GodloveD** [comment]
+1 on tests.  Will add tonight.  Also +1 on edits.  

BTW, your second test above seems a little weird right?  Is it a typo?  Did you mean something more like this?
```
$ SINGULARITYENV_PATH=/opt singularity exec docker://alpine /usr/bin/env | grep ^PATH=
PATH=/opt
```

**Comment by dtrudg** [comment]
If you just set /opt then the container won’t run - ‘env’ won’t be in PATH

Edit - see what you mean... fixing. Copy paste blunder.

**Review by dtrudg — APPROVED** [review]

**Review by ctmadison — APPROVED** [review]

**Comment by jakirkham** [comment]
Does this mean 2.5 is on it's way? :)

**Pull request merged**